### PR TITLE
[JENKINS-64107] Convert SSHD module to a plugin

### PR DESCRIPTION
**Pull request by kuisathaverat**
See [JENKINS-55582](https://issues.jenkins-ci.org/browse/JENKINS-55582) [JENKINS-64107](https://issues.jenkins-ci.org/browse/JENKINS-64107) [JENKINS-64104](https://issues.jenkins-ci.org/browse/JENKINS-64104).

### Proposed changelog entries

* Major RFE: Turn [SSHD module](https://github.com/jenkinsci/sshd-module) into a plugin, SSHD module is no longer bundle in Jenkins Core, it is provided as a plugin for further details see [JENKINS-55582](https://issues.jenkins-ci.org/browse/JENKINS-55582) and [JENKINS-64107](https://issues.jenkins-ci.org/browse/JENKINS-64107)

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [X] (If applicable) Jira issue is well described
- [X] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [X] Appropriate autotests or explanation to why this change has no tests
- ~~[ ] For dependency updates: links to external changelogs and, if possible, full diffs~~

### Desired reviewers

@jenkinsci/code-reviewers

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by jglick — COMMENTED** [review]
Will need to add a `test`-scoped dep to the `test` module since some functional tests in core do try to use `java -jar jenkins-cli.jar -ssh`.

File: test/pom.xml
Comment (jglick) [inline test/pom.xml:171]: You need to introduce a property in the root POM to define the version I think.

**Comment by kuisathaverat** [comment]
the publish stages failed due to the issues on the infra these last two days, all the tests and the ATH passes.

**Comment by jglick** [comment]
The JEP proposes inlining `ssh-cli-auth` but I do not see that here.

**Comment by kuisathaverat** [comment]
I was thinking of splitting the change into two PRs, but if you think I can include the ssh-cli-auth change here, I'd include it.

**Comment by jglick** [comment]
I think you have to. Otherwise we could release a split of a plugin before that plugin is actually ready.

**Comment by kuisathaverat** [comment]
@jglick I check the `ssh-cli-auth-module` and it depends on `sshd-module` for the tests, if I include the `ssh-cli-auth-module` as a dependency inside `sshd-module`, we will have a circular dependency for testing, so I wonder if we should move the code in `ssh-cli-auth-module` repo to the `sshd-module` repo and archive the `ssh-cli-auth-module` repo.

**Comment by jglick** [comment]
> move the code in `ssh-cli-auth-module` repo to the `sshd-module` repo and archive the `ssh-cli-auth-module` repo

That is precisely what I advocated in the JEP. Perhaps I did not word it clearly enough?

**Comment by oleg-nenashev** [comment]
+1 for having a single plugin BTW. I am a bit concerned about having no SSH CLI support on startup without a plugin, but I think we can live without it.

File: core/src/main/resources/jenkins/split-plugins.txt
Comment (kuisathaverat) [inline core/src/main/resources/jenkins/split-plugins.txt:33]:
```suggestion
sshd 2.265 3.0-rc164.cb4cccdfb4db
```

**Review by jglick — APPROVED** [review]
Approving conditional on upstream.

**Comment by kuisathaverat** [comment]
I have added the test on https://github.com/jenkinsci/sshd-module/pull/40 and update the PR with the master changes

What are the next steps? I guess we have to merge https://github.com/jenkinsci/sshd-module/pull/38 after this we can decide the final Jenkins Core version we will include this PR, and after that, we can finish and merge https://github.com/jenkinsci/sshd-module/pull/40

finally, if everything is fine we can test and merge https://github.com/jenkinsci/sshd-module/pull/37

File: core/src/main/resources/jenkins/split-plugins.txt
Comment (jglick) [inline core/src/main/resources/jenkins/split-plugins.txt:33]:
```suggestion
sshd 2.265 3.0
```

File: pom.xml
Comment (jglick) [inline pom.xml:118]:
```suggestion
    <sshd.plugin.version>3.0</sshd.plugin.version>
```

File: pom.xml
Comment (jglick) [inline pom.xml:172]: revert
```suggestion
        <!-- 
```

**Comment by timja** [comment]
test needs updating: `jenkins.plugins.DetachedPluginsUtilTest.checkJaxb`

File: core/src/main/resources/jenkins/split-plugins.txt
Comment (jglick) [inline core/src/main/resources/jenkins/split-plugins.txt:33]: Needs to be updated to reflect actual split version.

BTW @timja I should have mentioned in the PR to create the plugin that it would be appropriate to set the `jenkins.version` to an incremental version _after_ the last weekly, so it is not offered to anyone on the UC until the weekly with _this_ change is released.

Comment (timja) [reply core/src/main/resources/jenkins/split-plugins.txt:33]: right, easy to do quickly

Comment (timja) [inline core/src/main/resources/jenkins/split-plugins.txt:33]:
```suggestion
sshd 2.282 3.0.1
```

**Comment by jglick** [comment]
Ready to be taken out of draft status?

**Comment by timja** [comment]
> Ready to be taken out of draft status?

tests need a fix

File: pom.xml
Comment (jglick) [inline pom.xml:118]:
```suggestion
    <sshd.plugin.version>3.0.1</sshd.plugin.version>
```

?

Comment (jglick) [reply pom.xml:118]: BTW I cannot find the source tree for this update.

Comment (timja) [reply pom.xml:118]: not published yet as I need an incremental to set the jenkins.version in the plugin I believe?

Comment (timja) [reply pom.xml:118]: unless I just skip tests and publish with the next weekly?

Comment (jglick) [reply pom.xml:118]: Once https://github.com/jenkinsci/jenkins/runs/1961167481 finishes you should have something you can use.

**Comment by jglick** [comment]
https://github.com/jenkinsci/jenkins/blob/2ddf35821c119b7c219c2b0d7f113291f683f16c/test/src/test/java/jenkins/security/ClassFilterImplTest.java#L160 can be deleted

File: core/src/main/resources/jenkins/split-plugins.txt
Comment (jglick) [inline core/src/main/resources/jenkins/split-plugins.txt:33]:
```suggestion
sshd 2.281 3.0.1
```

see https://github.com/jenkinsci/jenkins/blob/1b000a59aa9eeb78f64487bfd4961a78cdcf64db/core/src/main/resources/jenkins/split-plugins.txt#L3-L6

**Comment by jglick** [comment]
Note that automated test coverage is not going to be of much help here. You need to take the actual generated `jenkins.war` and try to run it to see what happens.

**Comment by timja** [comment]
I tested this manually, setup wizard worked fine, but update center data for the plugin seems missing, cc @danielbeck 

```
wget http://updates.jenkins.io/current/update-center.actual.json
```

**Comment by timja** [comment]
https://github.com/jenkins-infra/update-center2/pull/484

**Review by timja — APPROVED** [review]

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by jglick** [comment]
Suggest https://github.com/jenkinsci/jenkins/pull/5049#issuecomment-784387631 though it is not a blocker.

File: core/src/main/resources/jenkins/split-plugins.txt
Comment (jglick) [inline core/src/main/resources/jenkins/split-plugins.txt:33]: I still do not see this on the UC?

Comment (timja) [reply core/src/main/resources/jenkins/split-plugins.txt:33]: ?

```
➜  wars cat update-center.actual.json | jq '.plugins["sshd"]'
{
  "buildDate": "Feb 23, 2021",
  "dependencies": [],
  "developers": [
    {
      "developerId": "timja"
    }
  ],
  "excerpt": "Adds SSH server functionality to Jenkins, exposing CLI commands through it",
  "gav": "org.jenkins-ci.modules:sshd:3.0.1",
  "labels": [],
  "minimumJavaVersion": "1.8",
  "name": "sshd",
  "popularity": 0,
  "previousTimestamp": "2021-02-22T19:33:24.00Z",
  "previousVersion": "3.0",
  "releaseTimestamp": "2021-02-23T15:53:28.00Z",
  "requiredCore": "2.282-rc30905.2ddf35821c11",
  "scm": "https://github.com/jenkinsci/sshd-plugin",
  "sha1": "xX2ALuJ47hl9sGVwrc6WptJTj+g=",
  "sha256": "Jmpl9bTjReS3qrwwTf3d0CTR8RyB/bZS3u+IQVzvEJg=",
  "title": "SSH server",
  "url": "https://updates.jenkins.io/download/plugins/sshd/3.0.1/sshd.hpi",
  "version": "3.0.1",
  "wiki": "https://plugins.jenkins.io/sshd"
}
```

Comment (jglick) [reply core/src/main/resources/jenkins/split-plugins.txt:33]:
```console
$ curl -sL 'https://updates.jenkins-ci.org/update-center.json?version=2.282' | head -n-1 | tail -n+2 | jq -r '.plugins | .["sshd"] | .version'
3.0
```

Comment (timja) [reply core/src/main/resources/jenkins/split-plugins.txt:33]: seems not portable command?

```
head: illegal line count -- -1
```

Comment (timja) [reply core/src/main/resources/jenkins/split-plugins.txt:33]: hmm weird, maybe update center is ignoring the version in the dynamic tiers, 'latest' has it

Comment (jglick) [reply core/src/main/resources/jenkins/split-plugins.txt:33]: Maybe does not refresh until after 2.282 actually comes out. If so, fine.

**Comment by daniel-beck** [comment]
Why are we considering this a split/detached plugin, when the functionality wasn't in core before?

It makes some sense from a configuration/functionality POV (but seems separate enough that functionality would easily be restored after installing the plugin, similar to https://plugins.jenkins.io/cctray-xml/ ), but needlessly adds a dependency from plugins with slightly older core baselines.

**Comment by kuisathaverat** [comment]
I think that the main reason to consider it a detached plugin is that is a feature you do not want to be installed. the use of Apache Mina SSH library in it is another good reason, it forces all plugins to use a version highly attached to the Jenkins Core, it difficult to bump that library version. As a plugin allows more flexibility to change the library version and helps other plugins to depend on a plugin and not on a Jenkins Core.

**Comment by daniel-beck** [comment]
Well, yeah, but those dependencies on plugins need to be explicit. Adding it to the split-plugins file doesn't accomplish that, because that only matters on cores newer than what the downstream plugin was developed against.

> I think that the main reason to consider it a detached plugin is that is a feature you do not want to be installed.

Well, the alternative is to just split it off and add a changelog entry, like we did for `cctray-xml`. Otherwise it'll just be installed automatically all the time because tons of plugins (almost all of them at this point) would get it as an implied dependency.

**Comment by kuisathaverat** [comment]
Oh!! I get you now, I simply put it in detached for backward compatibility as I did with the trilead plugin, to avoid friction.

**Comment by daniel-beck** [comment]
Well, trilead was on the core classpath, this one was not (AFAIUI).

**Comment by oleg-nenashev** [comment]
IIUC upgrade guideline is not required here simnce the plugin is detached, may have been confused by the changelog entry. Am I right or not?

**Comment by kuisathaverat** [comment]
yes, it is detached so nothing to do on the user side.

**Pull request merged**
